Add alert dialog

mrlem opened the pull request:
Add [Adw.AlertDialog](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.7/class.AlertDialog.html) component, also add a test that matches the one in the doc.

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Dialog.kt:
```diff
@@ -168,4 +170,66 @@ fun AboutDialog(
+@Composable
+fun AlertDialog(
+    modifier: Modifier = Modifier,
+    contentHeight: Int = 0,
+    contentWidth: Int = 0,
+    followsContentSize: Boolean = true,
```
These parameters have the wrong default values. It should be `-1` for the width/height (could you also change it on the other Dialog functions), and `false` for `followsContentSize`

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Dialog.kt:
```diff
@@ -168,4 +170,66 @@ fun AboutDialog(
     )
+}
+
+data class Response(
```
Question: would it be worth to:
 - add the "onChosen" callback here
 - remove the id
 - on the "real" onResponse callback, select the corresponding response, and invoke its callback
 - remove the "closeResponse" parameter, and let the user handle any action by using the existing `onClose` parameter

What do you think?

mrlem replied:
I agree we should be able to offer a higher level interface. I'm not fond, however, of mixing response + callback, because in compose a typical way to structure the app is to expose a state flow, on callbacks (both can typically originate from a viewmodel), which becomes hard if you mix state with callbacks.

I'll try to find a compromise!

mrlem replied:
I added a commit, with all other comments handled.
And about the responses, here what I did:
* remove the notion of id from the callbacks / parameters (it's still present in the Response data class though, as an id is required internally)
* handle the close action behind the scene to avoid letting the user deal with it
* didn't include a callback in the Response to avoid mixing state/callback
* renamed Response to AlertDialogResponse so the imports a clearer (adw.Response wasn't clear enough)

Does that look better to you?

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Dialog.kt:
```diff
@@ -168,4 +170,66 @@ fun AboutDialog(
+    closeResponse: String = "close",
+    defaultResponse: String? = null,
+    onResponse: (responseId: String) -> Unit,
+    content: @Composable () -> Unit = {},
```
It seems that passing this content will override the default alert dialog content, including the header/body/buttons etc.

I suggest we remove this parameter.

Optionally, we could have another function that has only the content parameter, instead of the various other options, in case the user wants to craft their own dialog.

mrlem replied:
Then again, if a user wants to create his own dialog, he/she can just use the `Dialog()` composable.

MMarco94 commented:
Nice addition, thanks! 

I've left some comments

mrlem commented:
I added a commit, with all other comments handled.
And about the responses, here what I did:

* remove the notion of id from the callbacks / parameters (it's still present in the Response data class though, as an id is required internally)
* handle the close action behind the scene to avoid letting the user deal with it
* didn't include a callback in the Response to avoid mixing state/callback
* renamed Response to `AlertDialogResponse` so the imports a clearer (adw.Response wasn't clear enough)

Does that look better to you?

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Dialog.kt:
```diff
@@ -168,4 +170,64 @@ fun AboutDialog(
+        updater = {
+            set(heading) { this.widget.heading = it }
+            set(body) { this.widget.body = it }
+            set(responses) {
```
We should also handle the case where the label text changed.

Calling `setResponseLabel` should do the trick

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Dialog.kt:
```diff
@@ -168,4 +170,64 @@ fun AboutDialog(
+)
+
+@Composable
+fun AlertDialog(
```
Minor: I think mandatory (non-default) parameters should go first, to make it easier to call this function without named arguments.

Since other functions are like that, I'll probably do a pass and fix it everywhere. When I have time, I'll try to set up detekt and see if it can enforce it automatically.

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Dialog.kt:
```diff
@@ -168,4 +170,64 @@ fun AboutDialog(
+    contentWidth: Int = -1,
+    followsContentSize: Boolean = false,
+    presentationMode: DialogPresentationMode = DialogPresentationMode.AUTO,
+    title: String? = null,
```
I tried to pass a title here, and it seems like it does nothing. Maybe we should remove this parameter

mrlem replied:
Indeed seems this inherited attribute is ignored in alert dialogs.

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Dialog.kt:
```diff
@@ -168,4 +170,64 @@ fun AboutDialog(
+    followsContentSize: Boolean = false,
+    presentationMode: DialogPresentationMode = DialogPresentationMode.AUTO,
+    title: String? = null,
+    onClose: () -> Unit = {},
```
Since this (and the other dialogs) don't close on their own, the caller almost certainly will pass a callback to `onClose`. 

Should we remove the default value? What do you think?

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Dialog.kt:
```diff
@@ -168,4 +170,64 @@ fun AboutDialog(
+        updater = {
+            set(heading) { this.widget.heading = it }
+            set(body) { this.widget.body = it }
+            set(responses) {
```
We should also remove responses that are no longer in the list

MMarco94 replied:
You could save this list of responses, and call `removeResponse` for the ones that have been removed

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Dialog.kt:
```diff
@@ -168,4 +170,64 @@ fun AboutDialog(
+    defaultResponse: AlertDialogResponse? = null,
+    onResponse: (AlertDialogResponse) -> Unit,
+) {
+    var connection: SignalConnection<*>? by remember { mutableStateOf(null) }
```
Minor: we should probably add a check to make sure `defaultResponse` is in `responses`

mrlem replied:
Added a log if not

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Dialog.kt:
```diff
@@ -168,4 +170,64 @@ fun AboutDialog(
+                connection?.disconnect()
+                connection = this.widget.onResponse(null) { responseId ->
+                    if (responseId == "close") return@onResponse
+                    responses.firstOrNull { it.id == responseId }
```
Minor, non-blocking: Should we be more ambitious, and use `single` instead of `firstOrNull`?

GTK generally fails open on errors, so maybe you're right, and ignoring errors is the right approach. Should we at least log a warning?

MMarco94 replied:
PS: I've added a logging framework on https://github.com/MMarco94/compose-4-gtk/commit/5f028f05cecf2b8bf5ee4de4e9536990c2a0ac25#diff-1ae823fcd1cfb976da8b4032656efc82f19a7abafb4f6ef9271a680d6dd3f9c5 if you want to see how to add a log

mrlem replied:
Left lenient behavior, but added a log.

MMarco94 commented:
I left a lot of comments, but the overall approach looks perfect. Thanks!

mrlem commented:
Rebased to get the new logging API, and took all remarks in account, except one: removing obsolete responses.
Not sure how to do this: there's no API to get existing responses for an alert dialog. I could probably use some remember trick to get value N-1 and guess from that which values disappear, but I'm not sure: if this is the best approach.

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Dialog.kt:
```diff
@@ -168,4 +173,76 @@ fun AboutDialog(
+                }
+            }
+            set(defaultResponse) {
+                if (it != null && responses.none { response -> response.id == it.id }) {
```
Nitpick: this check could be done at the very beginning of the function body (line 199), so it's easier for people looking at the source code to understand the requirement of the function. Moreover, this check won't run if `defaultResponse` is removed from `responses`, as this function won't be called again.

Also, I'd use a `requirement` here, and just crash: in this specific case, the caller is in complete control of the parameters, so there's no "normal" circumstance in which this would be the right behavior.

MMarco94 reviewed (changes requested):
I think the only important item remaining is to update the responses as they change.

mrlem commented:
* rebased PR on new sub composition implementation (nice change!).
* handled obsolete responses changes (tested it locally)
* did not move default response test outside the remember function (as this could lead to many useless occurrences of this message), but added the "responses" variable to the remember (because indeed there was an issue when responses changed)
* also added "responses" variable to the remember(onResponse), because otherwise it keeps an outdated version of the responses in the inner callback.

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Dialog.kt:
```diff
@@ -163,4 +167,84 @@ fun AboutDialog(
+            if (dialog.hasResponse(response.id)) {
+                dialog.setResponseLabel(response.id, response.label)
+            } else {
+                dialog.addResponse(response.id, response.label)
```
This one is a bit minor, but this approach won't preserve the order of the responses. 
For example, if a new response is added as the first item in the `responses` list, it will be added as the last button.

Maybe the right approach is, if any response was added/removed, to remove all buttons, and add them all again. 

Sorry for not noticing this in the last revision

mrlem replied:
Used this approach, your point was perfectly valid.
Also used a require as suggested.

MMarco94 reviewed (changes requested):
I think the approach is perfect. There's just a small issue to fix.

Thanks!

mrlem commented:
Used this approach, your point was perfectly valid.
Also used a require as suggested.

MMarco94 commented:
Thanks for the change!
